src/core/manager.c

```diff
@@ -2986,10 +2995,49 @@ void manager_check_finished(Manager *m) {
         manager_invalidate_startup_units(m);
 }
 
+static bool generator_path_any(const char* const* paths) {
+        char **path;
+        bool found = false;
+
+        /* Optimize by skipping the whole process by not creating output directories
+         * if no generators are found. */
+        STRV_FOREACH(path, (char**) paths)
+                if (access(*path, F_OK) == 0)
+                        found = true;
+                else if (errno != ENOENT)
+                        log_warning_errno(errno, "Failed to open generator directory %s: %m", *path);
+
+        return found;
+}
+
+static const char* environment_generator_binary_paths[] = {
+        "/run/systemd/environment-generators",
+        "/etc/systemd/environment-generators",
+        "/usr/local/lib/systemd/environment-generators",
+        ENVIRONMENT_GENERATOR_PATH,
+        NULL
+};
+
+static int manager_run_environment_generators(Manager *m) {
+        char **tmp = NULL; /* this is only used in the forked process, no cleanup here */
+        const gather_stdout_callbacks cb = {
+                gather_environment_one, &tmp,
+                gather_environment_two, &tmp,
+                gather_environment_three, &m->environment,
+        };
+
+        if (!generator_path_any(environment_generator_binary_paths))
+                return 0;
```

Comment:
There's a duality: environment.d dirs are only for users, but the generators themselves run for all manager instances, and are supposed to just do nothing in cases where they should do nothing. My motivation for this was that: a) it's quite possible to want to set the same set of variables for all environments, and having install two generators would be a bit more work, b) it's fairly easy to check if running as a user, c) doing just one dir simplifies the implementation and documentation.

But on second thought, it's not so obvious: in environment-d-generator, I used `getuid`, but it seems that environment.d directories should be also supported for the root's --user manager. In this case, that check is wrong. Instead, we should check if $USER is set.

I you think that there should be two separate sets of generators, I can change that. I still think that simply checking if $USER is set is simpler, but I don't feel particularly strongly about that.